R/epiparameter.R

```diff
@@ -0,0 +1,996 @@
+    epi_name = epi_name,
+    prob_distribution = prob_distribution,
+    uncertainty = uncertainty,
+    summary_stats = summary_stats,
+    auto_calc_params = auto_calc_params,
+    citation = citation,
+    metadata = metadata,
+    method_assess = method_assess,
+    notes = notes,
+    ...
+  )
+
+  # uncertainty is checked after new_epiparameter to use <epiparameter> methods
+  epiparameter$uncertainty <- .clean_uncertainty(
+    epiparameter,
+    prob_distribution = prob_distribution,
+    uncertainty_missing = missing(uncertainty)
+  )
+
+  param_names <- names(get_parameters(epiparameter)) %||% NA_character_
+
+  stopifnot(
+    "uncertainty must be provided for each parameter" =
+      length(param_names) == length(epiparameter$uncertainty),
+    "parameters and uncertainty must be named and match" =
+      identical(param_names, names(epiparameter$uncertainty)) ||
+      is.na(param_names)
+  )
+
+  # call epiparameter validator
+  assert_epiparameter(epiparameter)
+
+  # return epiparameter object
+  epiparameter
+}
+
+#' Assert an object is a valid `<epiparameter>` object
+#'
+#' @param x An \R object.
+#'
+#' @return Invisibly returns an `<epiparameter>`. Called for side-effects
+#' (errors when invalid `<epiparameter>` object is provided).
+#'
+#' @export
+assert_epiparameter <- function(x) {
+  if (!is_epiparameter(x)) {
+    stop("Object should be of class epiparameter", call. = FALSE)
+  }
+
+  list_names <- c(
+    "disease", "pathogen", "epi_name", "prob_distribution", "uncertainty",
+    "summary_stats", "citation", "metadata", "method_assess", "notes"
+  )
+  missing_list_names <- list_names[!list_names %in% attributes(x)$names]
+  if (length(missing_list_names) != 0) {
+    stop(
+      "Object is missing ", toString(missing_list_names), call. = FALSE
+    )
+  }
+
+  # check for class invariants
+  stopifnot(
+    "epiparameter must contain a disease (single character string)" =
+    checkmate::test_string(x$disease),
+    "epiparameter must contain an epidemiological distribution" =
+      checkmate::test_string(x$epi_name),
+    "epiparameter must contain a <distribution> or <distcrete> or NA" =
+    checkmate::test_multi_class(
+      x$prob_distribution, classes = c("distribution", "distcrete")
+    ) || checkmate::test_string(x$prob_distribution, na.ok = TRUE),
+    "epidisit must contain uncertainty, summary stats and metadata" =
+      all(
+        is.list(x$uncertainty), is.list(x$summary_stats), is.list(x$metadata)
+      ),
+    "epiparameter must contain a citation" =
+      inherits(x$citation, "bibentry"),
+    "epiparameter notes must be a character string" =
+      checkmate::test_string(x$notes)
+  )
+
+  invisible(x)
+}
+
+#' Test whether an object is a valid `<epiparameter>` object
+#'
+#' @param x An \R object.
+#'
+#' @return A boolean `logical` whether the object is a valid `<epiparameter>`
+#' object.
+#' @export
+test_epiparameter <- function(x) { # nolint cyclocomp_linter
```

Comment:
I like having the `lintr` check the cyclomatic complexity as I feel it does help me not to write overly complex functions. The default is 15 (https://lintr.r-lib.org/reference/cyclocomp_linter.html) which I think I'll leave the same for now. As you can see I don't comply with the cyclocomp linter too strictly by occasionally using `# nolint cyclocomp_linter`.

I think the best approach would either be to comply with the linter more strictly and remove the `# nolint` flags or use them sparingly like the current approach. I think all other Epiverse-TRACE packages use the default linter so it would be good to try and stay consistent across the organisation.